File: mced/src/generators/python/DigitalGeometryActions.mjs
export function defineDigitalGeometryActionsGenerators(pythonGenerator) {

    pythonGenerator.forBlock['digitalgeometryactions_create_digital_cube'] = function(block, generator) {
        const center = generator.valueToCode(block, 'center', pythonGenerator.ORDER_ATOMIC) || 'None';
const side_length = generator.valueToCode(block, 'side_length', pythonGenerator.ORDER_ATOMIC) || 'None';
const block_type = generator.valueToCode(block, 'block_type', pythonGenerator.ORDER_ATOMIC) || 'None';
const rotation_matrix = generator.valueToCode(block, 'rotation_matrix', pythonGenerator.ORDER_ATOMIC) || 'None';
        const code = `DigitalGeometryActions.create_digital_cube(${center}, ${side_length}, ${block_type}, ${rotation_matrix})\n`;
        return block.outputConnection ? [code.trim(), pythonGenerator.ORDER_ATOMIC] : code;
    };

    pythonGenerator.forBlock['digitalgeometryactions_create_digital_line'] = function(block, generator) {
        const point1 = generator.valueToCode(block, 'point1', pythonGenerator.ORDER_ATOMIC) || 'None';
const point2 = generator.valueToCode(block, 'point2', pythonGenerator.ORDER_ATOMIC) || 'None';
const block_type = generator.valueToCode(block, 'block_type', pythonGenerator.ORDER_ATOMIC) || 'None';
        const code = `DigitalGeometryActions.create_digital_line(${point1}, ${point2}, ${block_type})\n`;
        return block.outputConnection ? [code.trim(), pythonGenerator.ORDER_ATOMIC] : code;
    };

    pythonGenerator.forBlock['digitalgeometryactions_create_digital_sphere'] = function(block, generator) {
        const center = generator.valueToCode(block, 'center', pythonGenerator.ORDER_ATOMIC) || 'None';
const radius = generator.valueToCode(block, 'radius', pythonGenerator.ORDER_ATOMIC) || 'None';
const block_type = generator.valueToCode(block, 'block_type', pythonGenerator.ORDER_ATOMIC) || 'None';
        const code = `DigitalGeometryActions.create_digital_sphere(${center}, ${radius}, ${block_type})\n`;
        return block.outputConnection ? [code.trim(), pythonGenerator.ORDER_ATOMIC] : code;
    };

    pythonGenerator.forBlock['digitalgeometryactions_create_digital_disc'] = function(block, generator) {
        const center = generator.valueToCode(block, 'center', pythonGenerator.ORDER_ATOMIC) || 'None';
const radius = generator.valueToCode(block, 'radius', pythonGenerator.ORDER_ATOMIC) || 'None';
const normal = generator.valueToCode(block, 'normal', pythonGenerator.ORDER_ATOMIC) || 'None';
const block_type = generator.valueToCode(block, 'block_type', pythonGenerator.ORDER_ATOMIC) || 'None';
        const code = `DigitalGeometryActions.create_digital_disc(${center}, ${radius}, ${normal}, ${block_type})\n`;
        return block.outputConnection ? [code.trim(), pythonGenerator.ORDER_ATOMIC] : code;
    };

    pythonGenerator.forBlock['digitalgeometryactions_create_digital_tube'] = function(block, generator) {
        const p1 = generator.valueToCode(block, 'p1', pythonGenerator.ORDER_ATOMIC) || 'None';
const p2 = generator.valueToCode(block, 'p2', pythonGenerator.ORDER_ATOMIC) || 'None';
const outer_thickness = generator.valueToCode(block, 'outer_thickness', pythonGenerator.ORDER_ATOMIC) || 'None';
const inner_thickness = generator.valueToCode(block, 'inner_thickness', pythonGenerator.ORDER_ATOMIC) || 'None';
const block_type = generator.valueToCode(block, 'block_type', pythonGenerator.ORDER_ATOMIC) || 'None';
        const code = `DigitalGeometryActions.create_digital_tube(${p1}, ${p2}, ${outer_thickness}, ${inner_thickness}, ${block_type})\n`;
        return block.outputConnection ? [code.trim(), pythonGenerator.ORDER_ATOMIC] : code;
    };

    pythonGenerator.forBlock['digitalgeometryactions_create_digital_plane'] = function(block, generator) {
        const normal = generator.valueToCode(block, 'normal', pythonGenerator.ORDER_ATOMIC) || 'None';
const point_on_plane = generator.valueToCode(block, 'point_on_plane', pythonGenerator.ORDER_ATOMIC) || 'None';
const width = generator.valueToCode(block, 'width', pythonGenerator.ORDER_ATOMIC) || 'None';
const height = generator.valueToCode(block, 'height', pythonGenerator.ORDER_ATOMIC) || 'None';
const block_type = generator.valueToCode(block, 'block_type', pythonGenerator.ORDER_ATOMIC) || 'None';
        const code = `DigitalGeometryActions.create_digital_plane(${normal}, ${point_on_plane}, ${width}, ${height}, ${block_type})\n`;
        return block.outputConnection ? [code.trim(), pythonGenerator.ORDER_ATOMIC] : code;
    };

    pythonGenerator.forBlock['digitalgeometryactions_create_digital_ball'] = function(block, generator) {
        const center = generator.valueToCode(block, 'center', pythonGenerator.ORDER_ATOMIC) || 'None';
const radius = generator.valueToCode(block, 'radius', pythonGenerator.ORDER_ATOMIC) || 'None';
const inner_radius = generator.valueToCode(block, 'inner_radius', pythonGenerator.ORDER_ATOMIC) || 'None';
const block_type = generator.valueToCode(block, 'block_type', pythonGenerator.ORDER_ATOMIC) || 'None';
        const code = `DigitalGeometryActions.create_digital_ball(${center}, ${radius}, ${inner_radius}, ${block_type})\n`;
        return block.outputConnection ? [code.trim(), pythonGenerator.ORDER_ATOMIC] : code;
    };
}

File: mcshell/digitalgeometryactions.py
from mcshell.mcactions_base import MCActionsBase
from blockapily import mced_block
from mcshell.constants import *

from mcshell.mcvoxel import (
    generate_digital_tube_coordinates,
    generate_digital_plane_coordinates,
    generate_digital_ball_coordinates,
    generate_digital_cube_coordinates,
    generate_digital_disc_coordinates,
    generate_digital_line_coordinates,
    generate_digital_sphere_coordinates
)


class DigitalGeometryActions(MCActionsBase):
    def __init__(self, mc_player_instance, delay_between_blocks=0.01):
        super().__init__(mc_player_instance, delay_between_blocks)

    @mced_block(
        label="Create Digital Cube",
        center={'label': 'Center'},
        side_length={'label': 'Side Length'},
        block_type={'label': 'Block Type'},
        rotation_matrix={'label': 'Rotation Matrix'}
    )
    def create_digital_cube(self, center: Vec3, side_length: float, block_type: 'Block',rotation_matrix: 'Matrix3'):
        coords = generate_digital_cube_coordinates(center=center.to_tuple(), side_length=float(side_length),rotation_matrix=rotation_matrix.matrix)
        self._place_blocks_from_coords(coords, block_type)

    @mced_block(
        label="Create Digital Line",
        point1={'label': 'Start Point'},
        point2={'label': 'End Point'},
        block_type={'label': 'Block Type'}
    )
    def create_digital_line(self, point1: Vec3, point2: Vec3, block_type: 'Block'):
        coords = generate_digital_line_coordinates(p1=point1.to_tuple(), p2=point2.to_tuple())
        self._place_blocks_from_coords(coords, block_type)

    @mced_block(
        label="Create Digital Sphere",
        center={'label': 'Center'},
        radius={'label': 'Radius'},
        block_type={'label': 'Block Type'}
    )
    def create_digital_sphere(self, center: Vec3, radius: float, block_type: 'Block'):
        coords = generate_digital_sphere_coordinates(center=center.to_tuple(), radius=float(radius))
        self._place_blocks_from_coords(coords, block_type)

    @mced_block(
        label="Create Digital Disc",
        center={'label': 'Center'},
        radius={'label': 'Radius'},
        normal={'label': 'Normal'},
        block_type={'label': 'Block Type'}
    )
    def create_digital_disc(self, center: Vec3, radius: float, normal: Vec3, block_type: 'Block'):
        # coords = generate_digital_disc_coordinates(center=center.to_tuple(), radius=float(radius), normal=normal.to_tuple())
        coords = generate_digital_disc_coordinates(normal=normal.to_tuple(),center_point=center.to_tuple(),outer_radius=radius)
        self._place_blocks_from_coords(coords, block_type)


    @mced_block(
        label="Create Digital Tube",
        p1={'label': 'Start Point'},
        p2={'label': 'End Point'},
        outer_thickness={'label': 'Outer Thickness (Radius)'},
        inner_thickness={'label': 'Inner Thickness (Radius)'},
        block_type={'label': 'Block Type'}
    )
    def create_digital_tube(self, p1: Vec3, p2: Vec3, outer_thickness: float, inner_thickness: float, block_type: 'Block'):
        coords = generate_digital_tube_coordinates(p1=p1.to_tuple(), p2=p2.to_tuple(), outer_thickness=float(outer_thickness), inner_thickness=float(inner_thickness))
        self._place_blocks_from_coords(coords, block_type)

    # TODO: Width and Height are not respected
    @mced_block(
        label="Create Digital Plane",
        normal={'label': 'Normal'},
        point_on_plane={'label': 'Point on Plane'},
        width={'label': 'Width'},
        height={'label': 'Height'},
        block_type={'label': 'Block Type'}
    )
    def create_digital_plane(self, normal: Vec3, point_on_plane: Vec3, width: float, height: float, block_type: 'Block'):
        outer_rect_dims = (float(width), float(height))
        coords = generate_digital_plane_coordinates(normal=normal.to_tuple(), point_on_plane=point_on_plane.to_tuple(), outer_rect_dims=outer_rect_dims)
        self._place_blocks_from_coords(coords, block_type)

    @mced_block(
        label="Create Digital Ball",
        center={'label': 'Center'},
        radius={'label': 'Outer Radius'},
        inner_radius={'label': 'Inner Radius'},
        block_type={'label': 'Block Type'}
    )
    def create_digital_ball(self, center: Vec3, radius: float, inner_radius: float, block_type: 'Block'):
        coords = generate_digital_ball_coordinates(center=center.to_tuple(), radius=float(radius), inner_radius=float(inner_radius))
        self._place_blocks_from_coords(coords, block_type)

    # TODO: broken 
    # @mced_block(
    #     label="Create Digital Tetrahedron",
    #     v1={'label': 'Vertex 1'},
    #     v2={'label': 'Vertex 2'},
    #     v3={'label': 'Vertex 3'},
    #     v4={'label': 'Vertex 4'},
    #     inner_offset_factor={'label': 'Inner Offset Factor'},
    #     block_type={'label': 'Block Type'}
    # )
    # def create_digital_tetrahedron(self, v1: Vec3, v2: Vec3, v3: Vec3, v4: Vec3, inner_offset_factor: float, block_type: 'Block'):
    #     vertices = [v1.to_tuple(), v2.to_tuple(), v3.to_tuple(), v4.to_tuple()]
    #     coords = generate_digital_tetrahedron_coordinates(vertices=vertices, inner_offset_factor=float(inner_offset_factor))
    #     self._place_blocks_from_coords(coords, block_type)

